epersongroups.md

```diff
@@ -0,0 +1,310 @@
+# EPerson groups Endpoints
+[Back to the list of all defined endpoints](endpoints.md)
+
+## Main Endpoint
+**GET /api/eperson/groups**
+
+## Single EPerson Group
+**GET /api/eperson/groups/<:uuid>**
+
+```json
+{
+  "id": "617cf46b-535c-42d5-9d22-327ce2eff6dc",
+  "uuid": "617cf46b-535c-42d5-9d22-327ce2eff6dc",
+  "name": "Administrator",
+  "handle": null,
+  "metadata": [
+    {
+      "key": "dc.title",
+      "value": "Administrator",
+      "language": null
+    }
+  ],
+  "permanent": true,
+  "type": "group",
+  "_links": {
+    "groups": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+    },
+    "epersons": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/epersons"
+    },
+    "self": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc"
+    }
+  },
+  "_embedded": {
+    "groups": {
+      "_embedded": {
+        "groups": []
+      },
+      "_links": {
+        "self": {
+          "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+        }
+      },
+      "page": {
+        "number": 0,
+        "size": 0,
+        "totalPages": 1,
+        "totalElements": 0
+      }
+    }
+  }
+}
+```
+
+## Create new EPerson Group
+
+**POST /api/eperson/groups**
+
+To create a new EPerson Group, perform a post with the JSON below to the epersons endpoint when logged in as admin.
+
+```json
+{
+  "metadata": [
+    {
+      "key": "dc.title",
+      "value": "New Group 1"
+    }
+  ]
+}
+```
+
+## Sub Groups in a single parent EPerson Group
+
+### Get Sub Groups in a single parent EPerson Group
+
+**GET /api/eperson/groups/<:uuid>/groups**
+
+Filter using parameter:
+* all : retrieve indirect groups as well
```

Comment:
I share the same doubt than @tdonohue 
The "all" filter should be replaced with a search method or a "projection" once implemented to declare with depth of embedding you want in the response

Reply:
When comparing it with `/api/core/communities/search/subCommunities?parent=<:uuid>`, doesn't that imply `/api/eperson/groups/<:uuid>/groups` without the parameter would also become a search?

Reply:
@benbosman : I'm less concerned with the `/api/eperson/groups/<:uuid>/groups` as there's a clear "membership" between the parent group and the sub group.  However, if you wanted to refactor that into a `/search` endpoint, I'm OK with that.  It'd likely end up being `/api/eperson/groups/search/subGroups?parent=<:uuid>` then.

My question about the purpose of `?all=true` still exists though. I don't understand what that param is doing and how it is expected to be used in the Angular UI.  So, it's difficult for me to recommend exactly how we might rethink this endpoint.

Reply:
@tdonohue It seems I lost the primary use case for the `?all=true` in the refactoring at https://github.com/DSpace/Rest7Contract/pull/41/commits/c614a94ecfc983862a02ff3fa79e7690e4baadac
The previous REST Contract contained an option to get all the groups of an EPerson, e.g. similar to what is included on the profile page. Having the ability to see all groups an EPerson belongs to is frequently useful. I'll need to recreate the Contract to get the groups of an EPerson.
I agree the `?all=true` parameter would indeed make more sense as a search.

I would also like to know whether you see a substantial difference between a community and sub communities (implemented as a search) vs a group and sub groups (currently in the contract without a search). I just want to know whether there's a particular difference between both cases requiring them to be treated differently, or whether your opinion is that in both cases, both approaches would be good solutions and there's no need for them to be consistent.

Reply:
Hi @benbosman : since we are already talking about the need to have a `/search` method for the `?all=true` use case, it seems reasonable to also implement the subgroup endpoint as `/api/eperson/groups/search/subGroups?parent=<:uuid>`. 

As you noted, it also keeps this new endpoint consistent with the subcommunities endpoint.

To be honest, I had not noticed this inconsistency between subgroups and subcommunities until you pointed it out.  But, I agree with you that it's probably best to align them.

Reply:
we should have both. The main purpose of the /api/eperson/groups/:uuid/groups endpoint is to manage the association. We need to support DELETE, POST and optionally PUT to this endpoint to give the ability to add groups as subgroups.

Using the LinkRest annotation with a linkMethod we should be able to avoid code duplication and just use the same method for the search and for the link, see https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java#L717

Reply:
Implementing both a search and a direct get to retrieve subgroups doesn't make any sense either. The data should be retrievable using a single endpoint.

We should either:
* Remain consistent with the current subcommunities functionality
  * This would use `/api/eperson/groups/search/subGroups?parent=<:uuid>` as suggested by @tdonohue 
  * This doesn't need to impact DELETE and POST on `/api/eperson/groups/:uuid/groups` as long as it's clear this endpoint doesn't support GET (similar to `DELETE /api/core/items/<:uuid>/mappedCollections/<:collection_uuid>` not being supported as a GET in https://github.com/DSpace/Rest7Contract/blob/master/items.md#mapped-collections)
* Update the subcommunities endpoint
  * The https://github.com/DSpace/Rest7Contract/blob/master/communities.md#subcommunities-1 endpoint should be updated to use `/api/core/communities/<:uuid>/subcommunities`. And future potentials DELETE and POST would also happen on `/api/core/communities/<:uuid>/subcommunities`

Reply:
Looking back on this discussion, I'm having a hard time providing a direction/path forward.

I'd still like to understand the `?all=true` filter better...If that is needed, then I think it's *definitely* a search (and not a filter -- a filter should limit results, and not *expand* the number of results).

So, at a minimum, I think we need a `/api/eperson/groups/search/subGroups` endpoint to support the concept of `all=true`.  However, that does **not** require that we remove all GET support on `/api/eperson/groups/:uuid/groups`, instead we'd just need to document that this endpoint *only* returns subgroups that are *directly members of* the given group.  We can then point folks at the `/api/eperson/groups/search/subGroups` endpoint if they want to perform a different type of search within groups.

While alignment with Communities/SubCommunities is "nice", I don't think it's 100% required here.  Relationships between Communities & SubCommunities are quite different from relationships between Groups. Keep in mind, the Group <-> Group association is many-to-many, while Community <-> Community is *one-to-many* (a subcommunity never belongs to multiple parents, but a Community can have many subcommunities).  So, we shouldn't be treating these concepts as 100% identical... general alignment is nice, but not required.